### PR TITLE
Refactored features, extractor classes; streamlined data proc and model training; changed API

### PR DESCRIPTION
**Pull request by bdewilde**
First things first: There are A LOT of changes in this PR, and I'm not quite done, so many apologies in advance for this incomplete monster.

Here's the rundown:

- I refactored all features so that they're `sklearn`-style transformers, with key methods of `.fit()` and `.transform()`. The underlying logic that converts sequences of blocks into 2D arrays of numeric features has been moved into and optimized in cython (with the exception of CSS features, which rely entirely on regexes). There are a number of benefits of switching to `Transformer` classes: easy combining, parallelizing of feature generation plus joint parameter optimization via `sklearn.pipeline.FeatureUnion`, interoperability with other `sklearn` features such as `StandardScaler` for standardizing feature values by mean and standard deviation, plus nice inherited behaviors from the `sklearn` base classes such as pretty-printed instances. I moved them into their own subpackage, to be tidy.
- I refactored the core `ContentExtractionModel` into an `sklearn`-style classifier `Extractor` with key methods of `.fit()` and `.predict()`, plus the primary entry point for users of `.extract(html, encoding)`. This class flexibly combines features, a classifier model (`ExtraTreesClassifier` by default), and a couple other optimizable parameters in a way that is also `sklearn`-friendly and `pickle`-friendly. It's easy to use grid search to jointly optimize features (and their individual params), the classifier model (and its params), as well as block weighting and the prediction probability threshold.
- I added top-level convenience functions to extract content, comments, and content+comments using pre-trained and -optimized `Extractor` models. They are loaded lazily from disk once, then cached. Loading is very fast via `joblib` (bundled within `sklearn`), although the pickled, compressed model files are surprisingly large (~14MB apiece).
- I streamlined much of the data processing and model training code, e.g. by removing seemingly deprecated or unnecessary code, generalizing some functionality, and reducing the amount of hard-coding throughout. I took advantage of `sklearn`'s existing metrics for, e.g. accuracy and F1 score, and applied them to block-level binary classification performance. I also added jaccard similarity and the existing damerau levenshtein to evaluate extracted vs. gold standard text content. I removed plotting functionality because it seemed like a thing best left to developers, but could totally put diagnostic plots back in.
- I cleaned up pep8 violations, improved documentation in `sphinx`-compatible Google docstring style, and laid a bit more groundwork for Py3 compatibility. On that note: I did _not_ touch any `Blockifier`-related code, because it's so complicated and because you have other packages (?) that rely on it as-is. `blocks.pyx` seems to be the last bastion of Py3 incompatibility — in unicode vs. bytes issues, of course. It's also still the slowest part of extracting content from html, by far.

There are still a few to-do's:
- I've not finished porting tests over; temporarily, I'm skipping the two problematic test cases in unittest.
- There's still a bit of a bytes and unicode mishmash, particularly in the data processing code. Nothing appears to be _broken_, but I'd feel much better working entirely in unicode under the hood. This will probably wait until `blocks.pyx` gets sorted out, which I probably can't do on my own. I tried reading [this](http://cython.readthedocs.io/en/latest/src/tutorial/strings.html) but got horribly lost.
- I've not settled on final extractor models (although pretty good pickled model files are included in this PR) because performance characteristics are slightly different. I'm optimizing by block-level classification performance of F-beta score (e.g. F1) using `sklearn`'s grid search + cross-validation.
- I've not yet done a full and thorough runtime and classification performance comparison. Initial tests have shown feature generation to be about twice as fast (maybe owing to the built-in parallelization of `FeatureUnion`). Also, I've found my best models to have higher recall but lower precision than previous versions (see below), but as I said, I'm still tweaking things. Would appreciate your feedback on this!
- The "public" API has changed, as has the workflow for training new models. Documentation has not yet been changed, pending your feedback. Btw, I'm happy to work towards better backwards compatibility, if you think that's important.

perf-comparison
(purple is the old model, blue is the new model)

Apologies again for this huge PR, one change kept leading into another... Let me know what you think about all this.

### CODE REVIEW

**Comment by matt-peters** [comment]
Whoa, this is a big PR.. I'll hopefully have time to look at it later in the week.  One immediate thought without looking at the code is that a complete change of the API is likely a no-go for Moz since we have a fair amount of legacy code depending on the current API, warts and all.  To give you some idea about the difficulty of changing the API completely, Moz has four different projects in prod that depend on dragnet across different teams so it'd be a rather large effort to swap it out completely and one that is unlikely to get support.

**Comment by bdewilde** [comment]
@matt-peters I thought you'd said only the blockifier API was relied upon elsewhere — my bad. For what it's worth, the other projects could simply pin their dependency version to dragnet==1.1.0 until they felt like porting over, but since I don't know how entangled everything is, I don't know how much work the eventual porting would be. I'll see what I can do to emulate the old API. Just to be clear: Are they only using the basic API, e.g.

```
from dragnet import content_extractor
content_extractor.analyze(html_document)
```

?

**Comment by matt-peters** [comment]
Sorry if I indicated earlier that the blockifier was the only interface used.  Unfortunately it's impossible for me to give a complete list as I'm no longer at Moz and don't have access to the legacy code (perhaps @dlecocq or @sistawendy could add some more detail?).  So this is just from memory, but it likely includes:

* the blockfier interface, `BlockifyError`, lower level details of the `Blocks` class, the dependence on `lxml` as the HTML parser
* `content_extractor`, `content_comments_extractor`, `content_and_content_comments_extractor`: the `analyze` method (with various options, which were add at some point for a use case).  It's also possible that we are using `analyze_from_blocks` or the various other lower level methods in these classes too as a common pattern is to move the `lxml` parsing from dragnet elsewhere but re-use the parse tree through the lower level methods.

As far as I'm concerned all the data processing and model training stuff isn't important to maintain in it's current form (as long as functionality is maintained somehow) since this is all done offline and not at run time.

**Comment by bdewilde** [comment]
Hey @matt-peters , just checking in. I was waiting to hear back re: API changes before doing any more work on this, but: I'll try to do and submit a full comparison of extraction performance and runtime between this and the previous version sometime this week.

**Comment by matt-peters** [comment]
Sorry about the silence on this one.  I'm having a hard time reviewing this due to the massive size.  Is there any way to split this into smaller PRs?

Also, we'll get you an answer about potential API changes soon.

**Comment by bdewilde** [comment]
Hi @matt-peters , let me see what I can do about splitting this into smaller PRs... Changes to the features, and then the extractor, arose from changes I'd wanted to make to the model training. So, maybe I'll start from the bottom and work my way up rather than the other way round. If it's all the same to you, I'll leave this PR as-is and submit others piece-by-piece.

**Comment by matt-peters** [comment]
OK we've decided to pin internal Moz dependencies on the current pip version so can move forward with API changes.  I'll hopefully have a chance to review the API changes later today so we can at least fix those before breaking it into pieces and merging.

**Comment by nehalecky** [comment]
hey all. We're (@pakelley) also interested in helping port this library to Python 3. In trying to find a roadmap or any existing efforts, came across this PR. What's the status of this? If the idea with the effort is to still break it apart, we'd like to contribute where we can. Appreciate it. 😄

**Comment by bdewilde** [comment]
Hi @nehalecky , I've been totally diverted by other projects, so haven't yet had a chance to revisit this PR. Help would definitely be appreciated! These changes arose from my wanting a better API for training / applying the extraction model, and opting to follow `scikit-learn`'s lead. Code and documentation cleanup, as well as steps toward Py3-compatibility, were just bonuses.

FWIW I think `blocks.pyx` is the biggest blocker to making `dragnet` Py3-compatible.

**Comment by nehalecky** [comment]
Hi @bdewilde, really appreciate the quick response and the details. I also commented on a collaborative effort for the porting in #26 to see what engagement we can get there. Look forward to the next steps, soon. :)

**Comment by b4hand** [comment]
Closed in favor of #52.
